### ISSUE TITLE
Can't reuse filter in the same stream

### ISSUE THREAD

**Issue by jgoux** [sindresorhus/gulp-filter#32]
Hello,
I want to use a filter twice in a stream but I have an error the second time it's called : 
```
stream.js:94
      throw er; // Unhandled stream error in pipe.
            ^
Error: write after end
    at writeAfterEnd (/home/jgx/Projects/javascript/cqst2/node_modules/gulp-filter/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:144:12)
    at DestroyableTransform.Writable.write (/home/jgx/Projects/javascript/cqst2/node_modules/gulp-filter/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:192:5)
    at DestroyableTransform.ondata (stream.js:51:26)
    at DestroyableTransform.emit (events.js:95:17)
    at DestroyableTransform.<anonymous> (/home/jgx/Projects/javascript/cqst2/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:786:14)
    at DestroyableTransform.emit (events.js:92:17)
    at emitReadable_ (/home/jgx/Projects/javascript/cqst2/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:448:10)
    at emitReadable (/home/jgx/Projects/javascript/cqst2/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:444:5)
    at readableAddChunk (/home/jgx/Projects/javascript/cqst2/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:187:9)
    at DestroyableTransform.Readable.push (/home/jgx/Projects/javascript/cqst2/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:149:10)
```

And here is my stream : 
```javascript
return gulp.src(config.src)
    .pipe(rename({suffix: '.min'}))
    .pipe(sass(
      {
        sourcemapPath: config.sourcemapDest,
        style: 'compressed'
      }
    ))
    .pipe(cssFilter)
    .pipe(autoprefixer('last 1 version'))
    .pipe(csso())
    .pipe(cssFilter.restore())
    .pipe(gulp.dest(config.dest))
    // Can't reuse cssFilter, filter('**/*.css') works
    .pipe(cssFilter)
    .pipe(browserSync.reload({stream: true}));
```

As I said in the comment, if I replace the second `cssFilter` with `filter('**/*.css')` it works well.

**Comment by sindresorhus** [comment]
@nfroidure Thoughts?

**Comment by nfroidure** [comment]
By admitting this is an important feature, i think that the factory API is better, not a big fan of "reactivating" a filter which seems like an object re-instanciation.

On the other side, the benefit provided by this feature is mainly the fact to have only one reference (as in @adam-lynch proposition)  and a factory API wouldn't help for that matter:

```js
var cssFilter1 = filter('*.css'), cssFilter2;
return gulp.src(config.src)
    .pipe(...)
    .pipe(cssFilter1)
    .pipe(...)
    .pipe(cssFilter1.restore())
    .pipe(...)
    .pipe(cssFilter2 = cssFilter.clone()) // re-initialized the stream or something
    .pipe(...)
    .pipe(cssFilter2.restore())
    .pipe(...);
```

It not that better than:

```js
var cssMatch = '*.css';
var cssFilter1 = filter(cssMatch), cssFilter2 = filter(cssMatch)
return gulp.src(config.src)
    .pipe(...)
    .pipe(cssFilter1)
    .pipe(...)
    .pipe(cssFilter1.restore())
    .pipe(...)
    .pipe(cssFilter2) // re-initialized the stream or something
    .pipe(...)
    .pipe(cssFilter2.restore())
    .pipe(...);
```

Not sure it is that often you need to reuse a filter, also, not sure it is that good to be able to do so instead of maintaining 2 distinct refs for 2 distinct filters. Sometimes code expressiveness primes on DRY.

**Comment by nfroidure** [comment]
In fact, since we are considering breaking changes for a v2, another way to avoid adding extra references could be to do something like that
https://github.com/gulpjs/gulp/issues/905#issuecomment-74428764

So the filter function could accept a PassThrough stream in which to pass filtered files and restore himself in output.
```js
var filter = require('gulp-filter');
var multipipe = require('multipipe');

gulp.src('**')
  .pipe(filter('*.css', multipipe(cssTransformStream1, cssTransformStream2)))
  .pipe(gulp.dest('.'))

```

It would solve this issue since users would only have to create a reference for the passed in multimatch query to reuse the same filter multiple times.

**Comment by sindresorhus** [comment]
Should be fixed in 3.0.0: https://github.com/sindresorhus/gulp-filter/releases/tag/v3.0.0
